### ISSUE TITLE
[Clang] `-Wdangling` breaks when `-D_GLIBCXX_DEBUG` is used

### ISSUE THREAD

**Issue by dfrvfmst** [llvm/llvm-project#109627]
`_GLIBCXX_DEBUG` macro is used to enabled debug assertion in `libstdc++`. However, when it's enabled, `-Wdangling` no longer works in the following case:

https://godbolt.org/z/x3dcs8eb6

```cpp
#include <vector>

auto f()
{
#pragma GCC diagnostic push
#pragma GCC diagnostic error "-Wdangling"
    auto &&a = std::vector{1}.at(0);
#pragma GCC diagnostic pop
    return a;
}
```

The expected result:

```
error: object backing the pointer will be destroyed at the end of the full-expression [-Werror,-Wdangling-gsl]
```

The actual result: no compile error got triggered.

---

Not sure whether libstdc++ debug assertion also breaks other stuff. Was surprised to see clangd failed catching trivial dangling issue (and ASan got triggered instead) in one of my project when they were caught previously in other projects.

Clang version is 18.1.0 (as shown in the godbolt link).
